### PR TITLE
Td 4 reports

### PR DESCRIPTION
**Pull request by Apozzz**
5. Reports generation (“Sandėlio” programa)

a. Susipažinti su:

i. Spring scheduling task’ais

b. Sukurti scheduler’į kuris generuoja report’us, pvz tam tikro laiko periodo ataskaita su dalimi pardavimų informacijos.

c. Report’ą turėtų būti galima atsisiųsti CSV formatu (download endpoint’as)

### CODE REVIEW

File: warehouse/src/main/java/com/twoday/wms/warehouse/report/ReportFileServiceImpl.java [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+
+        try (FileWriter writer = new FileWriter(file)) {
+            writer.write(data);
+        } catch (IOException e) {
+            throw new ReportFileException(ERROR_WRITING_TO_FILE);
+        }
+    }
+
+    @Override
+    public File getFile(Optional<String> dateTime) {
```
Comment (ArnasPetraitis) [inline warehouse/src/main/java/com/twoday/wms/warehouse/report/ReportFileServiceImpl.java]: Optional is considered not good as function argument. find out why

File: warehouse/src/main/java/com/twoday/wms/warehouse/report/ReportScheduler.java [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+import com.twoday.wms.warehouse.report.interfaces.ReportFileService;
+import com.twoday.wms.warehouse.report.interfaces.ReportGeneratorService;
+
+import lombok.RequiredArgsConstructor;
+
+@Component
+@RequiredArgsConstructor
+public class ReportScheduler {
+
+    private static final String HOURLY_CRON_EXPRESSION = "0 0 * * * ?";
```
Comment (ArnasPetraitis) [inline warehouse/src/main/java/com/twoday/wms/warehouse/report/ReportScheduler.java]: cron should be in application properties

File: warehouse/src/main/java/com/twoday/wms/warehouse/report/constants/CsvConstants.java [hunk 3]
```diff
@@ -0,0 +1,16 @@
+package com.twoday.wms.warehouse.report.constants;
+
+
+public class CsvConstants {
+
+    private CsvConstants() {
+        throw new IllegalStateException("Utility class");
```
Comment (ArnasPetraitis) [inline warehouse/src/main/java/com/twoday/wms/warehouse/report/constants/CsvConstants.java:7]: if class is used only for values, you could change to interface

File: warehouse/src/main/java/com/twoday/wms/warehouse/report/interfaces/FieldAction.java [hunk 4]
```diff
@@ -0,0 +1,10 @@
+package com.twoday.wms.warehouse.report.interfaces;
```
Comment (ArnasPetraitis) [inline warehouse/src/main/java/com/twoday/wms/warehouse/report/interfaces/FieldAction.java]: should this be in `.report.` package... ? maybe it is not that coupled, and can live on its own?

File: warehouse/src/main/java/com/twoday/wms/warehouse/report/utils/ReflectionUtils.java [hunk 5]
```diff
@@ -0,0 +1,18 @@
...
+import java.lang.reflect.Method;
+
+public class ReflectionUtils {
+
+    private ReflectionUtils() {
+        throw new IllegalStateException("Utility class");
+    }
+    
+    public static Method findGetter(Field field) throws NoSuchMethodException {
+        String base = field.getName().substring(0, 1).toUpperCase() + field.getName().substring(1);
```
Comment (ArnasPetraitis) [inline warehouse/src/main/java/com/twoday/wms/warehouse/report/utils/ReflectionUtils.java]: there should already be an util for this :)

File: warehouse/src/main/java/com/twoday/wms/warehouse/purchase/Purchase.java [hunk 6]
```diff
@@ -24,18 +26,30 @@ public class Purchase {
     
     @Id
     @GeneratedValue
+    @CsvField(name = "Purchase ID", order = 1)
     private Long id;
 
     @ManyToOne(fetch = FetchType.LAZY)
     @JoinColumn(name = "user_id", nullable = false)
+    @CsvFields({
```
Comment (ArnasPetraitis) [inline warehouse/src/main/java/com/twoday/wms/warehouse/purchase/Purchase.java]: although nice implementation with annotation, this annotation in entity kind of violates single responsibility principle

Comment (ArnasPetraitis) [reply warehouse/src/main/java/com/twoday/wms/warehouse/purchase/Purchase.java]: maybe report generation from model could be separated by using dto's in report

**Review by ArnasPetraitis — CHANGES_REQUESTED** [review]
nice implementation with annotations, but usually annotations are created after multiple repeated cases. because it takes longer to implement and also YAGNI

File: warehouse/src/main/java/com/twoday/wms/warehouse/report/constants/CsvConstants.java (hunk 3 above)
Comment (Apozzz) [reply warehouse/src/main/java/com/twoday/wms/warehouse/report/constants/CsvConstants.java:7]: Gal, bet raso, kad tai labai blogai. 
image

**Review by ArnasPetraitis — APPROVED** [review]

File: warehouse/src/main/java/com/twoday/wms/warehouse/report/utils/ReflectionUtils.java (hunk 5 above)
Comment (Apozzz) [reply warehouse/src/main/java/com/twoday/wms/warehouse/report/utils/ReflectionUtils.java]: There is none. Unless BeansUtil, but the code then becomes bigger. Anyway i implemented it.

Comment (ArnasPetraitis) [reply warehouse/src/main/java/com/twoday/wms/warehouse/report/utils/ReflectionUtils.java]: https://docs.spring.io/spring-framework/docs/3.0.0.M1/javadoc-api/org/springframework/expression/spel/reflection/ReflectionUtils.html#findGetterForProperty(java.lang.String,%20java.lang.Class,%20boolean)

**Pull request merged**